icepyx/core/read.py

```diff
@@ -370,29 +252,7 @@ def __init__(
                 "Please use the `data_source` argument to specify your dataset instead."
             )
 
-        if isinstance(data_source, list):
-            # if data_source is a list pass that directly to _filelist
-            self._filelist = data_source
-        elif os.path.isdir(data_source):
-            # if data_source is a directory glob search the directory and assign to _filelist
-            data_source = os.path.join(data_source, "*")
-            self._filelist = glob.glob(data_source, **glob_kwargs)
-        elif isinstance(data_source, str):
-            if data_source.startswith("s3"):
-                # if the string is an s3 path put it in the _filelist without globbing
-                self._filelist = [data_source]
-            else:
-                # data_source is a globable string
-                self._filelist = glob.glob(data_source, **glob_kwargs)
-        else:
-            raise TypeError(
-                "data_source should be a list of files, a directory, the path to a file, "
-                "or a glob string."
-            )
-
-        # Remove any directories from the list (these get generated during recursive
-        # glob search)
-        self._filelist = [f for f in self._filelist if not os.path.isdir(f)]
+        self._filelist = _parse_source(data_source, glob_kwargs)
 
         # Create a dictionary of the products as read from the metadata
         product_dict = {}
```

Comment:
since it's not changed I can't call it out directly, but I was looking at line 294 (`elif len(all_products) == 0:`). From a coding perspective it makes sense to cover this case (possible list lengths we care about are 0, 1, >1). But, given line 275 (`self.is_s3 = self.is_s3[0]`) will fail with an index out of range error if the list length is 0, and the `is2ref.extract_product()` of line 267 will fail if the file isn't valid (i.e. doesn't contain a product), could we actually generate the case where `len(all_products)==0`? Is best coding practice to include the case anyway?

Reply:
This is a good catch. I think this is a remnant of when we used to take a product as a user argument and need to confirm that the found files match the product. Given that `is2ref.extract_product` will raise an error if no products are found I think it's safe to say that `product_dict` should never have 0 entries and this statement can be removed.

Reply:
I realized there is a second task that this if statement is performing, which is that it's the only place we actually check to make sure a non-zero number of files was given (or found by the glob statement). I think this is an important check, but leaving it here is a bit of an obfuscated way of checking. I added an if statement in the `_parse_source` function to do that instead.

Reply:
Good catch. I updated the test I had checking for an empty list to check for the error instead.